core/iwasm/common/wasm_memory.c

```diff
@@ -420,13 +420,28 @@ is_native_addr_in_shared_heap(WASMModuleInstanceCommon *module_inst,
                               uint8 *addr, uint32 bytes)
 {
     WASMSharedHeap *heap = get_shared_heap(module_inst);
+    uintptr_t base_addr = (uintptr_t)heap->base_addr;
+    uintptr_t addr_int = (uintptr_t)addr;
+    uintptr_t end_addr = addr_int + bytes;
 
-    if (heap && addr >= heap->base_addr
-        && addr + bytes <= heap->base_addr + heap->size
-        && addr + bytes > addr) {
-        return true;
+    if (!heap) {
+        return false;
     }
-    return false;
+
+    if (addr_int < base_addr) {
+        return false;
+    }
+
+    // Check for overflow
```

Comment:
had better use `/* .. */` to add the comment, and how about mentioning `integer overflow`?